builders/container-apps/startup-script.sh

```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+
+set -e
+
+# Wait for tarball to be placed in COMPRESSED_APP_LOCATION, and the type to be compressed data.
+# Waiting for the type to be compressed data also helps ensure that the file is fully uploaded before
+# proceeding.
+while [[ ! -f $COMPRESSED_APP_LOCATION || ! "$(file $COMPRESSED_APP_LOCATION)" =~ "compressed data" ]]
+do
+  echo "Waiting for app source to be uploaded to '$COMPRESSED_APP_LOCATION'..."
+  sleep 5
+done
+
+# Extract app code to CNB_APP_DIR directory.
+echo "Found app source at '$COMPRESSED_APP_LOCATION'. Extracting to $CNB_APP_DIR"
+mkdir -p $CNB_APP_DIR
+cd $CNB_APP_DIR
+tar -xzf "$COMPRESSED_APP_LOCATION"
+
+# Detecting runtime stack and setting correct oryx runtime image.
+runtime_stack=$(oryx dockerfile . | head -n 1 | sed 's/ARG RUNTIME=//')
+export CNB_RUN_IMAGE="mcr.microsoft.com/oryx/$runtime_stack"
```

Comment:
nit: any thoughts on a fallback plan we should have if this `oryx dockerfile` command fails or isn't able to detect a suitable platform for the given application?

Reply:
yeah this actually brings up a few good points. First, I know that we are planning on allowing customers to be able to manually override the runtime platform and version with some parameters that will be passed in, so that update should help customers not be blocked by oryx's failures to detect a language. As far as a default / fallback, I'm not quite sure. Node is probably the highest usage? So maybe that? Or we just hard fail and tell them to override the platform and version manually

Reply:
I'm in the boat that we should fail fast and notify the user to the best of our ability that we were unable to detect a suitable platform version for their application.

Although now that I'm thinking about it, that's something that we should be doing in the `detect` script of the actual buildpack, so the question then becomes: is there any difference in detection logic between `oryx detect` and `oryx dockerfile`? I'd assume no since they _should_ be using the same library, but I wonder if there's a difference between "we've detected a platform" (`oryx detect`) and "we've detected a specific platform version to use" (`oryx dockerfile`).

Reply:
So afaik, the `detect` script of the buildpack is purely for detecting which buildpacks to use, in case there are more than one buildpack, correct? So in our case where we only have a single buildpack, there really isn't any need for the `detect` phase to have any logic since it will ALWAYS choose the oryx buildpack that we made, right?

I've tried to implement it so that we are able to set the run image that is used within the `detect` phase of buildpack, but that didn't work which is why I needed to set it before we even run the buildpack phases.

But yeah, I agree about failing fast. Made https://devdiv.visualstudio.com/DevDiv/_boards/board/t/Oryx/Stories/?workitem=1775134 to finalize the capps builder (since its current state was just made for the demo)

Reply:
> So afaik, the detect script of the buildpack is purely for detecting which buildpacks to use, in case there are more than one buildpack, correct?

This is correct, but even though we only have a single buildpack in our builder, I think it would make sense to have that phase "fail" if the application provided by the user wouldn't be successfully built by our corresponding `build` script. I think it just comes down to semantics since we're OK either way, it's just a matter of where we want the "failed message" to pop up for the user. I'm honestly not sure if we have the ability to customize the message seen by the user if no compatible buildpack is found for the provided application source, so I think we're good with what you have now.